modal/_serialization.py

```diff
@@ -42,13 +46,142 @@ def persistent_load(self, pid):
             raise InvalidError("bad flag")
 
 
-def serialize(obj):
+def serialize(obj: Any) -> bytes:
     """Serializes object and replaces all references to the client class by a placeholder."""
     buf = io.BytesIO()
     Pickler(buf).dump(obj)
     return buf.getvalue()
 
 
-def deserialize(s: bytes, client):
+def deserialize(s: bytes, client) -> Any:
     """Deserializes object and replaces all client placeholders by self."""
     return Unpickler(client, io.BytesIO(s)).load()
+
+
+def _serialize_asgi(obj: Any) -> api_pb2.Asgi:
+    def flatten_headers(obj):
+        return [s for k, v in obj for s in (k, v)]
+
+    msg_type = obj.get("type")
+    if msg_type == "http":
+        return api_pb2.Asgi(
+            http=api_pb2.Asgi.Http(
+                http_version=obj["http_version"],
+                method=obj["method"],
+                scheme=obj.get("scheme"),
+                path=obj["path"],
+                query_string=obj["query_string"],
+                headers=flatten_headers(obj["headers"]),
+                client_host=obj["client"][0] if obj.get("client") else None,
+                client_port=obj["client"][1] if obj.get("client") else None,
+            )
+        )
+    elif msg_type == "http.request":
+        return api_pb2.Asgi(
+            http_request=api_pb2.Asgi.HttpRequest(
+                body=obj.get("body"),
+                more_body=obj.get("more_body"),
+            )
+        )
+    elif msg_type == "http.response.start":
+        return api_pb2.Asgi(
+            http_response_start=api_pb2.Asgi.HttpResponseStart(
+                status=obj["status"],
+                headers=flatten_headers(obj["headers"]),
+                trailers=obj.get("trailers"),
+            )
+        )
+    elif msg_type == "http.response.body":
+        return api_pb2.Asgi(
+            http_response_body=api_pb2.Asgi.HttpResponseBody(
+                body=obj.get("body"),
+                more_body=obj.get("more_body"),
+            )
+        )
+    elif msg_type == "http.response.trailers":
+        return api_pb2.Asgi(
+            http_response_trailers=api_pb2.Asgi.HttpResponseTrailers(
+                headers=flatten_headers(obj["headers"]),
+                more_trailers=obj.get("more_trailers"),
+            )
+        )
+    elif msg_type == "http.disconnect":
+        return api_pb2.Asgi(http_disconnect=api_pb2.Asgi.HttpDisconnect())
+    else:
+        logger.debug("skipping serialization of unknown ASGI message type %r", msg_type)
```

Comment:
If we get a message like a WebSocket response, or an unknown ASGI extension which somehow got enabled despite us not advertising support for it, we skip the serialization.

Reply:
would it be better to raise an exception? this will just ignore it effectively (people won't see the debug log unless they override the verbosity level)

Reply:
I think there's some kinds of messages we want to ignore as unimplemented without throwing (like the [lifespan protocol](https://asgi.readthedocs.io/en/latest/specs/lifespan.html)). But yeah for `websocket.*` events might be nice to throw explicitly

Reply:
I think we should ignore it, because there are many messages that we can't predict. The idea is "accept as many servers as possible, but require compliance on our side by users."

WebSockets just don't work on Modal, and that's fine documented as-is I think. We don't need to crash the user's app after a WebSocket request.

Reply:
Hmm fair enough — on second thought I don't think we actually forward WebSocket connection events on our side though, so it's a bit of a moot point?